# Elliptic Curve Cryptography (ECC)

Elliptic Curve Cryptography (ECC) leverages algebraic structure of elliptic curves over finite fields in order to generate public keys. It can solve the problem of performance-heavy operations with too large keys (f.ex. 2048-bit RSA), as usually for ECC 256 bits are used.

## 1 Elliptic curves

### 1.1 Elliptic curve: definition 

Elliptic curves are defined over a field $K$ and describe points in its cartesian product $K^2$; they are smooth and projective affine algebraic structures. 

The Weierstrass equation of an elliptic curve $E$ over a field $K$ is [[1]](#references): 

$$ E : y^2 + a_1xy + a_3y = x^3 + a_2x^2 + a_4x + a_6 $$

what in case of the fields with characteristic different than 2 or 3 (i.e. the smallest number of summands $n$ of the field's identity element $e$ so $ne=0$), the affinitic equation of the elliptic curve can be simplified to: 

$$ E : y^2 = x^3 + Ax + B $$ $$A, B \in K$$ 


### 1.2 Elliptic curve: point addition 

The structure of elliptic curves fulfills the necessary requirements for its points to form an algebraic group with a certain addition law. 

<div align="center">
<img width="500" src="assets/operations-on-ec.png" >
</div>
<div align="center">
Fig. 1.1: Operations on en elliptic curve [[2]](#references)
</div>

With $P, Q \in E$ - being two points on an elliptic curve group $E$ - and $L$ being the line connecting those points, the third point $R$ can be defined where the intersection of $L$ with $E$ happens. Then, the result of $P+Q$ is located on the residual point of the intersection $R'$. 

The coordinates of the result ($x_{res}$, $y_{res}$) can be calculated as:
$$x_{res} = s^2-x_p-x_q$$
$$y_{res} = -y_p+s(x_p-x_{res})$$
with $s$ being the slope of the line $L$ through $P$ and $Q$: 
$$s= y_p-y_q/x_p-x_q$$

With $O$ being the neutral point in the group $E$, the following properties are fullfiled: 
- $P + O = P$
- $P + Q = Q + P$
- if $P \in E$ then there exists a point $Q$ such that $P + Q = O$
- if $P, Q \in E$ then $P + Q \in E$
- for $P, Q, R \in E$ is satisfied $(P + Q) + R = P + (Q + R)$ 

### 1.4 Multiplication of EC point with integer

Any point $P$ on the elliptic curve can be added to itself resulting in: 
$$ P + P = 2 * P $$

Continuing the addition multiple $n$ times yeilds to: 

$$ P * n = P + P + ... + P (n-times)$$

As multiplication is defined by the point addition, it always results in another point on the elliptic curve. 

For point doubling, i.e. $R = 2*P = P + P$, the following formula holds: 
$$x_{R} = s^2-2x_p$$
$$y_{R} = -y_p+s(x_p-x_{res})$$
with $s$: 
$$s= (3x_p^2+a)/2y_p$$

### 1.5 Elliptic Curves over finite field

The ECs used in ECC are strictly defined over finite field $Fp$, with $p$ being a prime number bigger than 3. Hence, the field is a square matrix with dimensions $p \times p$. Hence, the points of a given curve are limited to integer coordinates within its field matrix and all algebraic operations will result in a point within that field.   



## 2 NIST elliptic curves

### 2.1 NIST Standard

National Institute of Standards and Technology (NIST) is an American national institute, working under the provisions of the Federal Information Security Management Act. It defines the standars of the curves and its parameters for different key-lengths, in order to achieve implementation with base security accepted by the USA government. As well, it outlines key-establishment schemes, key-derivation and key-pair managements methods, as well as recommendations for DLC-based (Discrete Logarithm Cryptography) key-agreement schemes[[4]](#references). 

For each key-length NIST provides two types of fields[[5]](#references): 
* **prime field**: it contains a prime number $p$ of elements, which are the integers modulo $p$, with the field arithmetic being implemented in terms of the arithmetic of integers modulo $p$;
* **binary field**: it contains $2^m$ elements (with $m$ being the degree of the field) which are bit strings of length $m$, and the field arithmetic is implemented in terms of operations on bits. 

As well, two types of curves are given: 
* **pseudorandom** curves: with coefficients that are generated from the output of the seeded hash function;
* **special** curves: with coefficients and the field being selected to optimize the efficiency of the arithmetic operations on the curve.

The special curves over prime fields are known as *Edwards* or *Montgomery* curves; meanwhile over binary field they are referred as *Koblitz* curves. 

In this paper, there will be inspected in detail psudorandom curves over prime fields. 

### 2.2 NIST pseudorandom curves over prime fields

#### 2.2.1 Curves parameters


Due to NIST official publication on *Recommended Elliptic Curves for Federal Government Use*[[6]](#references), all the pseudorandom curves over prime field curves are defined by: 

$$ E: y^2 = x^3 - 3x + b \mod p $$

with $a$ having a static value of $a=-3$ in every case for the reasons of efficiency. 

<div align="center">
<img width="600" align="center" src="assets/nist-basic-curve.png" />
</div>
<div align="center">
Fig. 2.1: Examples of curve E with different coefficients b, interactive access: <a href="https://www.desmos.com/calculator/shiz2q2qk4">desmos tool</a>
</div>

The following parameters are specifying the curves in details: 

* $p$: prime modulus 
* $n$: the order 
* $SEED$: the 160-bit long input seed to the domain parameter generation (SHA-1 algorithm)
* $c$: the output of the SHA-1 algorithm
* $b$: the coefficient of the curve, that grants satisfaction of condition $b^2c=-27 \mod p$
* $G$: the base point with two values: 
  * $Gx$: the $x$ coordinate of the base point
  * $Gy$: the $y$ coordinate of the base point

The parameters $p$ and $n$ are always given in decimal value, meanwhile the rest in hexadecimal. 

#### 2.2.2 P-256: 256-bit length curve


|     param     | value                                                                          |
| :-----------: | :----------------------------------------------------------------------------- |
|  $p_{(10)}$   | 115792089210356248762697446949407573530086143415290314195533631308867097853951 |
|  $n_{(10)}$   | 115792089210356248762697446949407573529996955224135760342422259061068512044369 |
| $SEED_{(16)}$ | c49d3608 86e70493 6a6678e1 139d26b7 819f7e90                                   |
|  $c_{(16)}$   | 7efba166 2985be94 03cb055c 75d4f7e0 ce8d84a9 c5114abc af317768 0104fa0d        |
|  $b_{(16)}$   | 5ac635d8 aa3a93e7 b3ebbd55 769886bc 651d06b0 cc53b0f6 3bce3c3e 27d2604b        |
|  $G_{x(16)}$  | 6b17d1f2 e12c4247 f8bce6e5 63a440f2 77037d81 2deb33a0 f4a13945 d898c296        |
|  $G_{y(16)}$  | 4fe342e2 fe1a7f9b 8ee7eb4a 7c0f9e16 2bce3357 6b315ece cbb64068 37bf51f5        |

#### 2.2.3 P-521: 512-bit length curve

|     param     | value                                                                                                                                                         |
| :-----------: | :------------------------------------------------------------------------------------------------------------------------------------------------------------ |
|  $p_{(10)}$   | 6864797660130609714981900799081393217269435300143305409394463459185543183397656052122559640661454554977296311391480858037121987999716643812574028291115057151 |
|  $n_{(10)}$   | 6864797660130609714981900799081393217269435300143305409394463459185543183397655394245057746333217197532963996371363321113864768612440380340372808892707005449 |
| $SEED_{(16)}$ | d09e8800 291cb853 96cc6717 393284aa a0da64ba                                                                                                                  |
|  $c_{(16)}$   | 0b4 8bfa5f42 0a349495 39d2bdfc 264eeeeb 077688e4 4fbf0ad8 f6d0edb3 7bd6b533 28100051 8e19f1b9 ffbe0fe9 ed8a3c22 00b8f875 e523868c 70c1e5bf 55bad637           |
|  $b_{(16)}$   | 051 953eb961 8e1c9a1f 929a21a0 b68540ee a2da725b 99b315f3 b8b48991 8ef109e1 56193951 ec7e937b 1652c0bd 3bb1bf07 3573df88 3d2c34f1 ef451fd4 6b503f00           |
|  $Gx_{(16)}$  | c6 858e06b7 0404e9cd 9e3ecb66 2395b442 9c648139 053fb521 f828af60 6b4d3dba a14b5e77 efe75928 fe1dc127 a2ffa8de 3348b3c1 856a429b f97e7e31 c2e5bd66            |
|  $Gy_{(16)}$  | 118 39296a78 9a3bc004 5c8a5fb4 2c7d1bd9 98f54449 579b4468 17afbd17 273e662c 97ee7299 5ef42640 c550b901 3fad0761 353c7086 a272c240 88be9476 9fd16650           |


## 3 Brainpool elliptic curves

### 3.1 Brainpool standard 

Brainpool elliptic curves are curves which are standardized by the Internet Engineering Task Force (IETF) - an international open standards organization - and are described in RFC-5639[[8]](#references).  

It presents a set of domain parameters defining cryptographically strong groups on elliptic curves for different key-length encryption schemes.

The curves are described with regard to Wierstrass' definition: 
$$ E : y^2 = x^3 + Ax + B $$

### 3.2 Brainpool curves 

### 3.3.1 Curves parameters 

The Brainpool curves are defined with similar parameters to NIST standard; used in the official publication symbols are: 

* $p$: prime number (modulus) specifying the base field, $p >3$ 
* $a$ and $b$: the coefficients of the $E$, satisfying $4a^3 + 27b^2 \mod p \ne 0$
* $G$: the base point with two values: 
  * $x$: the $x$ coordinate of the base point
  * $y$: the $y$ coordinate of the base point
* $q$: the prime order of the group generated by base point $G$ (equivalent of NIST $n$ parameter), such that $qG=0$
* $h$: the cofactor of $G$ in $E$, $h = E($Fp$)/q$

In the further implementations, the given here parameters naming convention will be used. 

### 3.3.2 brainpoolP256r1: 256-bit length curve

The curve for 256-bit length key is identified by the Curve-ID `brainpoolP256r1`, and has the following parameters:


|    param    | value                                                            |
| :---------: | :--------------------------------------------------------------- |
| $p_{(16)}$  | A9FB57DBA1EEA9BC3E660A909D838D726E3BF623D52620282013481D1F6E5377 |
| $q_{(16)}$  | A9FB57DBA1EEA9BC3E660A909D838D718C397AA3B561A6F7901E0E82974856A7 |
| $a_{(16)}$  | 7D5A0975FC2C3057EEF67530417AFFE7FB8055C126DC5C6CE94A4B44F330B5D9 |
| $b_{(16)}$  | 26DC5C6CE94A4B44F330B5D9BBD77CBF958416295CF7E1CE6BCCDC18FF8C07B6 |
| $Gx_{(16)}$ | 8BD2AEB9CB7E57CB2C4B482FFC81B7AFB9DE27E1E3BD23C23A4453BD9ACE3262 |
| $Gy_{(16)}$ | 547EF835C3DAC4FD97F8461A14611DC9C27745132DED8E545C1D54C72F046997 |
| $h_{(10)}$  | 1                                                                |


### 3.3.3 brainpoolP512r1: 512-bit length curve

The curve for 512-bit length key is identified by the Curve-ID `brainpoolP512r1`, and has the following parameters:

|    param    | value                                                                                                                            |
| :---------: | :------------------------------------------------------------------------------------------------------------------------------- |
| $p_{(16)}$  | AADD9DB8DBE9C48B3FD4E6AE33C9FC07CB308DB3B3C9D20ED6639CCA703308717D4D9B009BC66842AECDA12AE6A380E62881FF2F2D82C68528AA6056583A48F3 |
| $q_{(16)}$  | AADD9DB8DBE9C48B3FD4E6AE33C9FC07CB308DB3B3C9D20ED6639CCA70330870553E5C414CA92619418661197FAC10471DB1D381085DDADDB58796829CA90069 |
| $a_{(16)}$  | 7830A3318B603B89E2327145AC234CC594CBDD8D3DF91610A83441CAEA9863BC2DED5D5AA8253AA10A2EF1C98B9AC8B57F1117A72BF2C7B9E7C1AC4D77FC94CA |
| $b_{(16)}$  | 3DF91610A83441CAEA9863BC2DED5D5AA8253AA10A2EF1C98B9AC8B57F1117A72BF2C7B9E7C1AC4D77FC94CADC083E67984050B75EBAE5DD2809BD638016F723 |
| $Gx_{(16)}$ | 81AEE4BDD82ED9645A21322E9C4C6A9385ED9F70B5D916C1B43B62EEF4D0098EFF3B1F78E2D0D48D50D1687B93B97D5F7C6D5047406A5E688B352209BCB9F822 |
| $Gy_{(16)}$ | 7DDE385D566332ECC0EABFA9CF7822FDF209F70024A57B1AA000C55B881F8111B2DCDE494A5F485E5BCA4BD88A2763AED1CA2B2FA8F0540678CD1E0F3AD80892 |
| $h_{(10)}$  | 1                                                                                                                                |


## 4 Implementation: Elliptic curve over finite prime field and ECDH

### 4.1 Preeliminary considerations 

In order to achieve certain level of security, the selection of the curve parameters and the base point is important; hence, in this realization, the Brainpool defined curves will be used. 

Apart of that, the required protection depends on the implementation and its environment: if the hardware that runs it is completely secured and separated, the time-invariant implementation may be sufficient. If the implementation of ECC is located on the physically accessible device, it has to be robust against side-channel attacks (power consumption and its analysis, electromagnetic emanations, active fault injection, etc.), then more security measures needs to be taken, with techniques of randomization being essential in it[[3]](#references). In this lab, the hardware vurnerabilities are not taken in considerations, hence the focus will be on achieving time-constant implementation.

### 4.2 ECDH: Elliptic Curve Diffie-Hellman Key Exchange

The ECDH (Elliptic Curve Diffie-Hellman Key Exchange) is the scheme that allows derivation of the shared secret key over an insecure channel, based on the private-public key pairs of both sides. It follows the traditional Diffie–Hellman Key Exchange, nevertheless, the key pairs are of elliptic curve type and point multiplication is performed instead of the modular exponentiation.

Considering following properties: 
* $P_{base}$ - base point of the curve
* $A_{priv}$ - integer, private key of A 
* $A_{pub}$ - point on the curve, $A_{pub}=A*P_{base}$, public key of A 
* $B_{priv}$ - integer, private key of B 
* $B_{pub}$ - point on the curve, $B_{pub}=B*P_{base}$, public key of B
* $K$ - shared secret key, $K=A_{priv}*B_{priv}*P_{base}$

the algorithm executes following flow: 
```
┌──────────────────────────────────────────────────────────────┐
│              A                               B               │
│              │                               │               │
│    ┌─────────▼─────────┐           ┌─────────▼─────────┐     │
│    │   Choose A_priv   │           │   Choose B_priv   │     │
│    └─────────┬─────────┘           └─────────┬─────────┘     │
│              │                               │               │
│   ┌──────────▼───────────┐        ┌──────────▼───────────┐   │
│   │    Compute A_pub     │        │    Compute B_pub     │   │
│   │A_pub = A_priv*P_base │        │B_pub = B_priv*P_base │   │
│   └──────────┬───────────┘        └──────────┬───────────┘   │
│              │                               │               │
│     ┌────────▼─────────┐            ┌────────▼─────────┐     │
│     │  Exchange A_pub  └────────────►  Exchange B_pub  │     │
│     │    and B_pub     ◄────────────┐    and A_pub     │     │
│     └────────┬─────────┘            └────────┬─────────┘     │
│              │                               │               │
│      ┌───────▼─────────┐             ┌───────▼─────────┐     │
│      │   Compute K     │             │   Compute K     │     │
│      │  K=A_priv*B_pub │             │  K=B_priv*A_pub │     │
│      └─────────────────┘             └─────────────────┘     │
└──────────────────────────────────────────────────────────────┘
```

 
This results in both sides, A and B, having the same secret shared key that can be used for symmetric encryption. 

Based on that, the implementation will be divided into following steps: 
* curve data type definition 
* curve and field generation functions
* point addition function
* double-and-add algorithm for faster multiplication
* Diffie-Hellman algorithm for shared key generation

### 4.3 Definition of the curve parameters

The curve parameters were declared with use of Python's `dataclass`[[7]](#references), introduced in version 3.7. This allowed to keep the code cleaner and to approach further computations rather functionally than object oriented. 



In [438]:
from dataclasses import dataclass 
from typing import Callable

@dataclass
class Point:
    x: int 
    y: int
    def __eq__(self: Point, other: Point):
        return self.x == other.x and self.y == other.y
    def __neq__(self: Point, other: Point):
        return not self == other

@dataclass 
class Field: 
    p: int # prime modulus 
    q: int # order of the base point 
    h: int = 1

@dataclass
class EllipticCurve:
    a: int # coefficient
    b: int # coefficient
    base: Point # base point
    field: Field
    inf: Point = Point(0, 0) # point of infinity
    name: str = "undefined" 


In [381]:
from enum import Enum

# list of available curves 
class ELLIPTIC_CURVES(Enum):
    bp256 = 'brainpoolP256r1'
    bp512 = 'brainpoolP512r1'

brainpool_256 = EllipticCurve(
    name=ELLIPTIC_CURVES.bp256.value,
    a=0x7D5A0975FC2C3057EEF67530417AFFE7FB8055C126DC5C6CE94A4B44F330B5D9,
    b=0x26DC5C6CE94A4B44F330B5D9BBD77CBF958416295CF7E1CE6BCCDC18FF8C07B6,
    base=Point(
        x=0x8BD2AEB9CB7E57CB2C4B482FFC81B7AFB9DE27E1E3BD23C23A4453BD9ACE3262,
        y=0x547EF835C3DAC4FD97F8461A14611DC9C27745132DED8E545C1D54C72F046997
    ),
    field=Field(
        p=0xA9FB57DBA1EEA9BC3E660A909D838D726E3BF623D52620282013481D1F6E5377,
        q=0xA9FB57DBA1EEA9BC3E660A909D838D718C397AA3B561A6F7901E0E82974856A7,
        h=1
    )
)

brainpool_512 = EllipticCurve(
    name=ELLIPTIC_CURVES.bp512.value,
    a=0x7830A3318B603B89E2327145AC234CC594CBDD8D3DF91610A83441CAEA9863BC2DED5D5AA8253AA10A2EF1C98B9AC8B57F1117A72BF2C7B9E7C1AC4D77FC94CA,
    b=0x3DF91610A83441CAEA9863BC2DED5D5AA8253AA10A2EF1C98B9AC8B57F1117A72BF2C7B9E7C1AC4D77FC94CADC083E67984050B75EBAE5DD2809BD638016F723,
    base=Point(
        x=0x81AEE4BDD82ED9645A21322E9C4C6A9385ED9F70B5D916C1B43B62EEF4D0098EFF3B1F78E2D0D48D50D1687B93B97D5F7C6D5047406A5E688B352209BCB9F822,
        y=0x7DDE385D566332ECC0EABFA9CF7822FDF209F70024A57B1AA000C55B881F8111B2DCDE494A5F485E5BCA4BD88A2763AED1CA2B2FA8F0540678CD1E0F3AD80892
    ),
    field=Field(
        p=0xAADD9DB8DBE9C48B3FD4E6AE33C9FC07CB308DB3B3C9D20ED6639CCA703308717D4D9B009BC66842AECDA12AE6A380E62881FF2F2D82C68528AA6056583A48F3,
        q=0xAADD9DB8DBE9C48B3FD4E6AE33C9FC07CB308DB3B3C9D20ED6639CCA70330870553E5C414CA92619418661197FAC10471DB1D381085DDADDB58796829CA90069,
        h=1
    )
)

ELLIPTIC_CURVES_LIST = [brainpool_256, brainpool_512]

# function to retrieve a curve to work on 
getCurve: Callable[[ELLIPTIC_CURVES], EllipticCurve] = lambda name : next(filter(lambda ec: ec.name == name,  ELLIPTIC_CURVES_LIST))

In [382]:
print(getCurve(ELLIPTIC_CURVES.bp256.value))
print(getCurve(ELLIPTIC_CURVES.bp512.value))
print(getCurve('brainpoolP256r1'))


EllipticCurve(a=56698187605326110043627228396178346077120614539475214109386828188763884139993, b=17577232497321838841075697789794520262950426058923084567046852300633325438902, base=Point(x=63243729749562333355292243550312970334778175571054726587095381623627144114786, y=38218615093753523893122277964030810387585405539772602581557831887485717997975), field=Field(p=76884956397045344220809746629001649093037950200943055203735601445031516197751, q=76884956397045344220809746629001649092737531784414529538755519063063536359079, h=1), inf=Point(x=0, y=0), name='brainpoolP256r1')
EllipticCurve(a=6294860557973063227666421306476379324074715770622746227136910445450301914281276098027990968407983962691151853678563877834221834027439718238065725844264138, b=3245789008328967059274849584342077916531909009637501918328323668736179176583263496463525128488282611559800773506973771797764811498834995234341530862286627, base=Point(x=6792059140424575174435640431269195087843153390102521881468023012732047482579853077

In [460]:
from typing import List
# common divisor functions -> for natural positive numbers 
# returns list of divisors of number n
divs: Callable[[int], List[int]] = lambda n: [i for i in range(1, n + 1)  if n % i == 0]

# interface: GCD and Bezout coefficients (x, y), so ax+bx=gcd(a, b) 
@dataclass 
class GcdProps: 
    gcd: int # prime modulus 
    x: int # order of the base point 
    y: int = 1

egcdStep: Callable[[int, int, GcdProps], GcdProps] = lambda a, b, g: \
    GcdProps(g.gcd, g.y-(b//a)*g.x, g.x)

# extended GCD euclidean algorithm
egcd: Callable[[int, int], GcdProps] = lambda a, b: \
    GcdProps(b, 0, 1) if a == 0 else \
        egcdStep(a, b, egcd(b%a, a))

print(egcd(25, 100))
print(egcd(56, 63))
print(egcd(1308, 729))

# modular multiplicative inverse integer (a/b vs b/a); it is necessary to compute the slope f.ex.
# gcd(a, prime) = x*a + y*prime -> x is the inverse

        #None if egcd(i, ec.field.p).gcd != 1 else \

multInvert: Callable[[EllipticCurve, int], int] = lambda ec, i: \
    0 if i==0 else \
        (lambda g=egcd(i, ec.field.p): None if g.gcd != 1 else g.x%ec.field.p)()

print(10, multInvert(ec1, 10))


GcdProps(gcd=25, x=1, y=0)
GcdProps(gcd=7, x=-1, y=1)
GcdProps(gcd=3, x=34, y=-61)
10 <function <lambda>.<locals>.<lambda> at 0x7fbeb746a790>


In [386]:
# check if point is on curve [in the field Fp]
# y^2 % p = (x^3 + Ax + B) % p
# isOnCurve2: Callable[[Point, EllipticCurve], bool] = lambda p, ec: (p.y**2 % ec.field.p == (p.x**3 + ec.a*p.x + ec.b) % ec.field.p)

isOnCurve: Callable[[Point, EllipticCurve], bool] = lambda p, ec: \
    (p.y**2 - ((p.x**2 + ec.a)*p.x + ec.b) % ec.field.p) == 0

ec1 = getCurve(ELLIPTIC_CURVES.bp256.value)
G = ec1.base 
print(isOnCurve(G, ec1))

False


In [454]:
# point of infinnity check
isZero: Callable[[EllipticCurve, Point], bool] = lambda ec, p: (p.y == ec.inf.y and p.x == ec.inf.x)

# invert the point (y axis)
invert: Callable[[EllipticCurve, Point], Point] = lambda ec, p: \
    p if isZero(ec, p) else Point(p.x, (-p.y)%ec.field.p)

# check if two points are inverse 
areInverse: Callable[[EllipticCurve, Point, Point], bool] = lambda ec, p1, p2: \
    (p1 == invert(ec, p2))

Gi = invert(ec1, G)
print(areInverse(ec1, G, Gi))


True


In [377]:
# slope in the field
slope1: Callable[[EllipticCurve, Point, Point], float] = lambda ec, p1, p2: \
    ((p1.y-p2.y)%ec.field.p/(p1.x-p2.x)%ec.field.p)%ec.field.p if (p1.x - p2.x != 0) \
        else ((3*p1.x**2+2*ec.a)/(2*p1.y))%ec.field.p \
            if (isOnCurve(p1, ec) and isOnCurve(p2, ec)) \
                else ValueError('One of points is not on the curve')

# slope in the field
slope: Callable[[EllipticCurve, Point, Point], float] = lambda ec, p1, p2: \
    ((p1.y-p2.y)%ec.field.p/(p1.x-p2.x)%ec.field.p)%ec.field.p if (p1 != p2) \
        else ((3*p1.x**2+2*ec.a)/(2*p1.y))%ec.field.p 

# get x of the resulting point, given slope
sumX: Callable[[EllipticCurve, Point, Point, int], int] = lambda ec, p1, p2, s: \
    (((s**2)%ec.field.p)-p1.x-p2.x)%ec.field.p

# calculate y coordinate, given x, and return point 
sumY: Callable[[EllipticCurve, Point, int, int], Point] = lambda ec, p1, x, s: \
    Point(x, (-p1.y+s*(p1.x-x))%ec.field.p)

# add two points, given the slope
ecAdd: Callable[[EllipticCurve, Point, Point, int], Point] = lambda ec, p1, p2, s: \
    sumY(ec, p1, sumX(ec, p1, p2, s), s)
    
# get the slope and add two points
pointAddition: Callable[[EllipticCurve, Point, Point], Point] = lambda ec, p1, p2: \
    p1 if isZero(ec, p2) else p2 if isZero(ec, p1) else \
        ec.inf if areInverse(ec, p1, p2) else \
            ecAdd(ec, p1, p2, slope(ec, p1, p2)) 
    

print(pointAddition(ec1, G, G))

Point(x=6.941541091606376e+76, y=2.6474136143366915e+76)


In [378]:
# point doubling 
pointDouble: Callable[[EllipticCurve, Point], Point] = lambda ec, p1: \
    ecAdd(ec, p1, p1, slope(ec, p1, p1)) 

print(pointDouble(ec1, G))

Point(x=6.941541091606376e+76, y=2.6474136143366915e+76)


In [379]:
from functools import reduce
#point multiplication 
#n*P = P + P + P ... + P 

pointMultiply: Callable[[EllipticCurve, Point, int], Point] = lambda ec, p1, n: \
    reduce((lambda a, b: pointAddition(ec, a, b)), [p1 for _ in range(0, n)])
   # reduce(pointDouble(ec, p1), range(0, n)) #[print(x, ': ', p1) for x in range(0, n)]

def pointMultiply2(ec:EllipticCurve, p: Point, n: int):
    # if n > field.q (order) then nG might = 0 #
    print(p)
    sum = p
    for _ in range(0,n-1):
        print(isOnCurve(sum, ec))
        print(sum.y**2 % ec.field.p)
        print((sum.x**3 + ec.a*sum.x + ec.b) % ec.field.p)
        sum = pointAddition(ec, sum, p) 
        print(_, ': ', sum)
        #print(sum.x%ec.field.p - p.x%ec.field.p)
        #print(sum.y%ec.field.p, p.y%ec.field.p)

#print(pointMultiply(ec1, G, 5))
pointMultiply2(ec1, G, 10)


Point(x=63243729749562333355292243550312970334778175571054726587095381623627144114786, y=38218615093753523893122277964030810387585405539772602581557831887485717997975)
False
28548113553162845316177995980873749235278879389902401598904818955824094474059
28548113553162845316177995980873749235278879389902401598904818955824094474059
0 :  Point(x=6.941541091606376e+76, y=2.6474136143366915e+76)
False
3.2919732697587013e+76
1.4835899501907488e+76
1 :  Point(x=2.111077212846459e+76, y=2.205892982913026e+76)
False
4.489382552845488e+76
2.607282335579719e+76
2 :  Point(x=6.941541091606376e+76, y=5.482602656791508e+76)
False
6.890868447118318e+76
1.4835899501907488e+76
3 :  Point(x=2.111077212846459e+76, y=7.515719693537551e+76)
False
3.2941387781553177e+76
2.607282335579719e+76
4 :  Point(x=6.941541091606376e+76, y=1.7277594616698278e+75)
False
3.4972736245555734e+76
1.4835899501907488e+76
5 :  Point(x=2.111077212846459e+76, y=6.889778553469176e+75)
False
4.120022909054529e+75
2.607282335579719e

In [368]:
# slope in the field
s1: Callable[[EllipticCurve, Point, Point], float] = lambda ec, p1, p2: \
    ((p1.y-p2.y)/(p1.x-p2.x))%ec.field.p  

s2: Callable[[EllipticCurve, Point], float] = lambda ec, p: \
    ( (3*(p.x**2)+2*ec.a)/(2*p.y))%ec.field.p   


def pAdd(p1: Point, p2: Point, ec: EllipticCurve):
    if areInverse(ec, p1, p2): 
        print('inverse')
        return ec.inf
    if (p1.x == p2.x):
        if (p1.y == p2.y): 
            print('same')
            s=s2(ec, p1)
            print(s)
            print(s% ec.field.p)
            xr = ((s**2)% ec.field.p - p1.x - p2.x) % ec.field.p
            yr = (-p1.y - (s*(xr-p1.x))% ec.field.p ) % ec.field.p
            r = Point(xr, yr)
            print('res: ',r, isOnCurve(r, ec))
            print( ((((3*(p1.x**2 % ec.field.p)% ec.field.p)+(2*ec.a)% ec.field.p)% ec.field.p)/(2*p1.y% ec.field.p))% ec.field.p  )
            s= ((((3*(p1.x**2 % ec.field.p)% ec.field.p)+(2*ec.a)% ec.field.p)% ec.field.p)/(2*p1.y% ec.field.p))% ec.field.p 
            # calc doubling
            xr = ((s**2)% ec.field.p - p1.x - p2.x) % ec.field.p
            yr = (-p1.y - (s*(xr-p1.x))% ec.field.p ) % ec.field.p
            r = Point(xr, yr)
            print('res: ',r, isOnCurve(r, ec))
            return r
        else: 
            print('same x, diff y, lol') 
    else:
        print('diff')
        s=s1(ec, p1, p2)
        print(s) 
        # calc adding


pAdd(G, G, ec1)

same
3.2125946629652635e+75
3.2125946629652635e+75
res:  Point(x=6.941541091606376e+76, y=6.514047744665873e+76) False
0.34722540397952883
res:  Point(x=2.7282453294966015e+76, y=5.1153010047858536e+76) False


In [ ]:
def addPoints1(ec:EllipticCurve, p1: Point, p2: Point):
        y_diff = (P_2.y - P_1.y) % self.char
        x_diff = (P_2.x - P_1.x) % self.char
        # if one p is zero, return the other one
        if isZero(ec, p1):
            return p2
        if isZero(ec, p2):
            return p1
        # if inverse points 
        if areInverse(ec, p1, p2):
            return ec.inf
        # adding a point to itself
        if p1 == p2:
            # if no slope
            if p1.y == 0:
                return ec.inf
            slope = ((3*p1.x**2 + 2*ec.a*p1.x + ec.b) * mult_inv(2*P_1.y, self.char)) % self.char
        else:
            slope = (y_diff * mult_inv(x_diff, self.char)) % self.char
        nu = (P_1.y - ld*P_1.x) % self.char
        x = (ld*ld - self.a - P_1.x - P_2.x) % self.char
        y = (-ld*x - nu) % self.char
        return Point(x,y)


### 4.4 Double and add algorithm 

The *Montgomery ladder* approach computes the point multiplication in a fixed amount of time. This can prevent leveraging time or power-consumption measurements and their analysis in order to perform a side-channel attack. 

In [366]:
# double and add algorithm 

n = 16 
nb = bin(n)[2:]

print(n, nb)

16 10000


# References <a id='references'></a>

- [[1]](https://perso.univ-rennes1.fr/christophe.ritzenthaler/cours/elliptic-curve-course.pdf): Ritzenhaler C., "Introduction to elliptic curves", Université de Rennes, 2013-2014, online access: https://perso.univ-rennes1.fr/christophe.ritzenthaler/cours/elliptic-curve-course.pdf

- https://cryptobook.nakov.com/asymmetric-key-ciphers/elliptic-curve-cryptography-ecc

- [3] https://csrc.nist.gov/csrc/media/events/workshop-on-elliptic-curve-cryptography-standards/documents/papers/session4-merkle-johannes.pdf - Requirements for Elliptic Curves for High-Assurance Applications, chapter 3.5 Implementation Security

- [4] https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar2.pdf
- [5] https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-186-draft.pdf page 9
- [6] https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf Appendix D: Recommended Elliptic Curves for Federal
Government Use, p. 87

- [7] https://www.python.org/dev/peps/pep-0557/ dataclasses

- [8] https://www.rfc-editor.org/rfc/pdfrfc/rfc5639.txt.pdf -> Elliptic Curve Cryptography (ECC) Brainpool Standard Curves and Curve Generation, chapter 3 Domain Parameter Specification

- ? https://www.rfc-editor.org/rfc/rfc8734.html#section-appendix.a-1 brainpool test vectors for later

- https://www.bsi.bund.de/SharedDocs/Downloads/EN/BSI/Publications/TechGuidelines/TR03111/BSI-TR-03111_V-2-1_pdf.pdf?__blob=publicationFile&v=1 

https://billatnapier.medium.com/barebones-p256-1700ff5a4  

https://girlstalkmath.com/2018/07/02/elliptic-curve-cryptography-2/ multiplicative inverse in mod arithmetic

my ascii charts https://asciiflow.com/#/share/eJyrVspLzE1VssorzcnRUcpJrEwtUrJSqo5RqohRsrI0NdaJUaoEsowszYGsktSKEiAnRunRlD1DGsXE5AFJBYoAbY1xJKDNiThjCDvv0ZQmOhlFIFamEYgzFGN7Hk1pwIlAJuGWnTIBq9OApHNGfn5xqoJjfEFRZhmGjyG%2Bg6pxQqhpItmnhBCKpVPw%2BmUNXtkZtIxPwj4hFKOoviUQqYTjFRG7aM4Eme6cn1tQWgKO3NIkdP9DPQpT44SkpgnTOIgRttB0ohUQn5QITBGoxjlB1TihqcFiHFUQsemFYJJBJByalShUKgTwJhjSiwCIx1wrkjMS89Lh6YTYGJi2C0mvE0wvlqIBFn6JeSnI6Wx6CwkRDdLriJZGSQlkEsof%2FAlqAEsffF4kJRHhTkVkJCGor2DliDdWn0LrETQ12OIQotXbFlrQQJILVuO8baEFjSNUDS7jKE8QxBQxpCcHUsEgNWZIoxilWqVaACONwUA%3D)